### PR TITLE
Modular Arrow

### PR DESCRIPTION
**Pull request by raulraja**
The following PR splits the code in the following modules:
```
arrow-annotations
arrow-annotations-processor
arrow-core (Option, Either, Eval..)
arrow-data (Coproduct, ListKW, ...)
arrow-docs
arrow-effects
arrow-effects-kotlinx-coroutines
arrow-effects-rx2
arrow-effects-test
arrow-examples
arrow-free (Free, FreeAplicative, Yoneda...)
arrow-instances (instances for core and data)
arrow-kindedj
arrow-mtl (FunctorFilter, MonadReader ...)
arrow-optics
arrow-recursion
arrow-syntax (option, functor, applicative, ...)
arrow-test
arrow-typeclasses
arrow-validation
```
Relevant decreased jar sizes without transitive dependencies
```
arrow-annotations-0.4.0.jar | 22 KB (no deps)
arrow-core-0.4.0.jar | 145 KB (deps: annotations)
arrow-typeclasses-0.4.0.jar | 237 KB (deps: core)
arrow-data-0.4.0.jar | 533 KB (deps: core)
arrow-instances-0.4.0.jar | 1296 KB (deps: core, data, typeclasses)
arrow-free-0.4.0.jar | 237 KB (deps: data, instances)
arrow-syntax-0.4.0.jar | 1344 KB (deps: annotations, data, instances)
```

Includes as a bonus improved output for Ank: 

output

### CODE REVIEW

File: ank-core/src/main/kotlin/io/kategory/ank/formatter.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package io.kategory.ank
```
Comment (Guardiola31337) [inline ank-core/src/main/kotlin/io/kategory/ank/formatter.kt]: Why `formatter.kt` is under `io.kategory.ank` package? Shouldn't be `arrow.ank`?

File: ank-core/src/main/kotlin/arrow/ank/interpreter.kt [hunk 2]
```diff
@@ -123,46 +119,63 @@ fun extractCodeImpl(source: String, tree: ASTNode): ListKW<Snippet> {
...
 fun compileCodeImpl(snippets: Map<File, ListKW<Snippet>>, classpath: ListKW<String>): ListKW<CompiledMarkdown> {
-    println(":runAnk -> started compilation")
-    val ioBoundPool = Executors.newFixedThreadPool(snippets.size)
-    val IOPool = object : CoroutineDispatcher() {
-        override fun dispatch(context: CoroutineContext, block: Runnable) {
-            ioBoundPool.submit(block)
+    println(colored(ANSI_PURPLE, AnkHeader))
+    val sortedSnippets = snippets.toList()
+    val result = sortedSnippets.mapIndexed { n, (file, codeBlocks) ->
+        //pb.extraMessage = "${file.parentFile.name}/${file.name}"
```
Comment (pakoito) [inline ank-core/src/main/kotlin/arrow/ank/interpreter.kt]: 🔥

File: arrow-data/src/main/kotlin/arrow/data/Try.kt [hunk 3]
```diff
@@ -28,7 +25,7 @@ sealed class Try<out A> : TryKind<A> {
         tailrec fun <A, B> tailRecM(a: A, f: (A) -> TryKind<Either<A, B>>): Try<B> {
             val ev: Try<Either<A, B>> = f(a).ev()
             return when (ev) {
-                is Failure -> Try.monadError().raiseError(ev.exception)
+                is Failure -> monadError<TryHK, Throwable>().raiseError<B>(ev.exception).ev()
```
Comment (pakoito) [inline arrow-data/src/main/kotlin/arrow/data/Try.kt]: Can't we just call the RaiseError constructor directly?

**Comment by pakoito** [comment]
Ank's snippet progress is off by one.

https://travis-ci.org/kategory/arrow/builds/323092343#L6552

**Comment by raulraja** [comment]
@pakoito it isn't really off by 1, just showing progress bar only for files that have code fences. I have changed it so it shows all progress bars and fills the progress bars even if the files are empty.

**Pull request merged**